expertsystem/reaction/solving.py

```diff
@@ -83,14 +91,37 @@ class GraphSettings:
     node_settings: Dict[int, NodeSettings] = attr.ib(factory=dict)
 
 
+# a tuple of rules defines a group which together violates all possible
+# combinations that were processed
+def convert_to_names(
+    rules: Dict[int, Set[Tuple[Rule, ...]]]
+) -> Dict[int, Set[Tuple[str, ...]]]:
```

Comment:
Why convert to a `str`?

Reply:
This is related to the issue of uniquely identifying a rule. The user is only interested why a problem does not have any solutions, so for example because of c parity violation. A string is enough for this. Currently the rule objects are used themselves, and define a human readable str (repr) method as well as a hash function. This is not very transparent.